src/lib/charm/openstack/keystone_saml_mellon.py

```diff
@@ -101,17 +83,24 @@ def sp_private_key_file(self):
     def sp_location_config(self):
         return SP_LOCATION_CONFIG
 
+    def get_fid_data(self):
```

Comment:
Echoing @fnordahl below, I think this function is in the 'wrong' place.  By pulling endpoint/interface into a `ConfigurationAdapter` a linkage is being created between config and interfaces, which should only be done in the charm class (according to the design of `charms.openstack`).  It means that this `ConfigurationAdapter` class now has a endpoint side-effect that needs to be accommodated in testing, which isn't normally the case with a configuration adapter.

This function, and it's dependants, either needs to move to a `RelationAdapter` class for the endpoint (if it can be constructed only using endpoint/relation data for that endpoint), or into the charm class of the function needs to mix config and relation data.

Reply:
Everything has been moved now.